flink-formats/flink-protobuf/src/main/java/org/apache/flink/formats/protobuf/serialize/RowToProtoConverter.java

```diff
@@ -0,0 +1,111 @@
+
+/**
+ * {@link RowToProtoConverter} can convert flink row data to binary protobuf message data by codegen
+ * process.
+ */
+public class RowToProtoConverter {
+    private static final Logger LOG = LoggerFactory.getLogger(ProtoToRowConverter.class);
+
+    private final Method encodeMethod;
+
+    public RowToProtoConverter(RowType rowType, PbFormatConfig formatConfig)
+            throws PbCodegenException {
+        try {
+            Descriptors.Descriptor descriptor =
+                    PbFormatUtils.getDescriptor(formatConfig.getMessageClassName());
+            Class<?> messageClass = Class.forName(formatConfig.getMessageClassName());
+
+            PbCodegenAppender codegenAppender = new PbCodegenAppender();
+            String uuid = UUID.randomUUID().toString().replaceAll("\\-", "");
+            String generatedClassName = "GeneratedRowToProto_" + uuid;
+            String generatedPackageName = RowToProtoConverter.class.getPackage().getName();
+            codegenAppender.appendLine("package " + generatedPackageName);
+            codegenAppender.appendLine("import " + AbstractMessage.class.getName());
+            codegenAppender.appendLine("import " + Descriptors.class.getName());
+            codegenAppender.appendLine("import " + RowData.class.getName());
+            codegenAppender.appendLine("import " + ArrayData.class.getName());
+            codegenAppender.appendLine("import " + StringData.class.getName());
+            codegenAppender.appendLine("import " + ByteString.class.getName());
+            codegenAppender.appendLine("import " + List.class.getName());
+            codegenAppender.appendLine("import " + ArrayList.class.getName());
+            codegenAppender.appendLine("import " + Map.class.getName());
+            codegenAppender.appendLine("import " + HashMap.class.getName());
+
+            codegenAppender.appendSegment("public class " + generatedClassName + "{");
+            codegenAppender.appendSegment(
+                    "public static AbstractMessage "
+                            + PbConstant.GENERATED_ENCODE_METHOD
+                            + "(RowData rowData){");
+            codegenAppender.appendLine("AbstractMessage message = null");
+            PbCodegenSerializer codegenSer =
+                    PbCodegenSerializeFactory.getPbCodegenTopRowSer(
+                            descriptor, rowType, formatConfig);
+            String genCode = codegenSer.codegen("message", "rowData");
+            codegenAppender.appendSegment(genCode);
+            codegenAppender.appendLine("return message");
+            codegenAppender.appendSegment("}");
+            codegenAppender.appendSegment("}");
+
+            String printCode = codegenAppender.printWithLineNumber();
+            LOG.debug("Protobuf encode codegen: \n" + printCode);
+            Class generatedClass =
+                    PbCodegenUtils.compileClass(
+                            this.getClass().getClassLoader(),
```

Comment:
Does the class loader to be used here need to be a class loader that can be obtained from `getUserCodeClassLoader` in the `InitializationContext` argument of the `open` method of `PbRowDataSerializationSchema`? I think it is better to use `UserCodeClassLoader`.

Reply:
Thanks for your comment. What if I use this.getClass().getClassLoader() to load the generated class? What is the difference of InitializationContext.UserCodeClassLoader ?

Reply:
I haven't experienced any particular errors with this ProtoBuf formatter, but I was curious about it because I have experienced NoClassDefFoundError and ClassNotFoundException related to Flink's class loader in the past.
I'm not familiar with Flink's class loader. If the current implementation looks fine, I think it's fine to leave it as is.